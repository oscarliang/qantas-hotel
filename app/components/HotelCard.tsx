import React from 'react';
import Image from 'next/image';
import { StarIcon, CircleIcon } from './Icons';

type HotelCardProps = {
    id: string;
    propertyId: string;
    title: string;
    address: string[];
    imageUrl: string;
    imageCaption: string;
    rating: {
        ratingValue: number;
        ratingType: 'star' | 'self';
    };
    promotion?: {
        title: string;
        type: string;
    };
    roomName: string;
    price: {
        amount: number;
        currency: string;
    };
    savings?: {
        amount: number;
        currency: string;
    } | null;
    cancellationType: 'FREE_CANCELLATION' | 'NOT_REFUNDABLE';
};

const HotelCard: React.FC<HotelCardProps> = ({
    title,
    address,
    imageUrl,
    imageCaption,
    rating,
    promotion,
    roomName,
    price,
    savings,
    cancellationType,
}) => {
    const renderRating = () => {
        const icons = [];
        const fullStars = Math.floor(rating.ratingValue);
        const halfStar = rating.ratingValue % 1 !== 0;

        const IconComponent = rating.ratingType === 'star' ? StarIcon : CircleIcon;

        for (let i = 0; i < fullStars; i++) {
            icons.push(<IconComponent key={i} filled={true} />);
        }

        if (halfStar) {
            icons.push(<IconComponent key="half" filled={false} />);
        }

        // Add empty icons to reach 5 in total
        const emptyIcons = 5 - Math.ceil(rating.ratingValue);
        for (let i = 0; i < emptyIcons; i++) {
            icons.push(<IconComponent key={`empty-${i}`} filled={false} />);
        }

        return <div className="flex my-1 sm:mt-2 sm:ml-2">{icons}</div>;
    };

    return (
        <div className="flex flex-col sm:flex-row  border-b border-gray-200 py-4">
            {/* Promotion Tag */}
            <div className="relative">
                {promotion && (
                    <span className="absolute top-2 left-0 bg-white text-red-600 text-xs py-1 px-2 z-10">
                        {promotion.title}
                    </span>
                )}

                {/* Hotel Image */}
                <Image
                    src={imageUrl}
                    alt={imageCaption}
                    width={145}
                    height={125}
                    className="object-cover"
                />
            </div>

            {/* Hotel Details */}
            <div className="flex-1 px-2 sm:px-4 py-2">
                <div className="flex flex-col sm:flex-row justify-between">
                    <div className="relative min-h-32 sm:min-h-28">
                        <div className="flex flex-col sm:flex-row justify-content align-middle">
                            <h2 className="text-lg font-bold">{title}</h2>
                            {/* Rating */}
                            {renderRating()}
                        </div>
                        <address className="text-sm text-gray-600 not-italic">
                            {address.join(', ')}
                        </address>

                        {/* Room Type */}
                        <p className="text-red-600 text-sm mt-2 font-medium underline">{roomName}</p>

                        {/* Cancellation */}
                        {cancellationType === 'FREE_CANCELLATION' && (
                            <p className="text-sm text-green-600 mt-2 absolute bottom-0">Free cancellation</p>
                        )}
                    </div>

                    {/* Price Information */}
                    <div className="text-right mt-4 sm:mt-0">
                        <div className="text-sm text-gray-600">1 night total (AUD)</div>
                        <div className="text-2xl font-bold mt-1">
                            <span className="text-lg">$</span>
                            {price.amount.toFixed(0)}
                        </div>

                        {/* Savings */}
                        {savings && (
                            <div className="text-red-600 text-sm">
                                Save ${savings.amount.toFixed(0)}~
                            </div>
                        )}
                    </div>
                </div>
            </div>
        </div>
    );
};

export default HotelCard; 