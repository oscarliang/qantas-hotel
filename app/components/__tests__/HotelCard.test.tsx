import React from 'react';
import { render, screen } from '@testing-library/react';
import HotelCard from '../HotelCard';

// Mock the Icons component
jest.mock('../Icons', () => ({
    StarIcon: ({ filled }: { filled: boolean }) => (
        <span data-testid={`star-icon-${filled ? 'filled' : 'empty'}`}>★</span>
    ),
    CircleIcon: ({ filled }: { filled: boolean }) => (
        <span data-testid={`circle-icon-${filled ? 'filled' : 'empty'}`}>○</span>
    ),
}));

describe('HotelCard Component', () => {
    const defaultProps = {
        id: 'test-id',
        propertyId: 'P12345',
        title: 'Test Hotel',
        address: ['123 Test St', 'Test City'],
        imageUrl: '/test-image.jpg',
        imageCaption: 'Test Hotel Image',
        rating: {
            ratingValue: 4.5,
            ratingType: 'star' as const,
        },
        roomName: 'Deluxe Room',
        price: {
            amount: 199.99,
            currency: 'AUD',
        },
        savings: {
            amount: 20,
            currency: 'AUD',
        },
        cancellationType: 'FREE_CANCELLATION' as const,
    };

    it('renders hotel details correctly', () => {
        const { container } = render(<HotelCard {...defaultProps} />);

        expect(screen.getByText('Test Hotel')).toBeInTheDocument();
        expect(screen.getByText('123 Test St, Test City')).toBeInTheDocument();
        expect(screen.getByText('Deluxe Room')).toBeInTheDocument();
        expect(screen.getByText(/\$\s*\d+/)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('renders star rating correctly', () => {
        render(<HotelCard {...defaultProps} />);

        // There should be 4 filled stars and 1 half star for a 4.5 rating
        const filledIcons = screen.getAllByTestId('star-icon-filled');
        expect(filledIcons).toHaveLength(4);

        const emptyIcons = screen.getAllByTestId('star-icon-empty');
        expect(emptyIcons).toHaveLength(1);
    });

    it('renders circle rating when type is self', () => {
        render(
            <HotelCard
                {...defaultProps}
                rating={{
                    ratingValue: 4,
                    ratingType: 'self'
                }}
            />
        );

        // Should render circle icons instead of stars
        const filledIcons = screen.getAllByTestId('circle-icon-filled');
        expect(filledIcons).toHaveLength(4);

        const emptyIcons = screen.getAllByTestId('circle-icon-empty');
        expect(emptyIcons).toHaveLength(1);
    });

    it('shows free cancellation when available', () => {
        render(<HotelCard {...defaultProps} />);

        expect(screen.getByText('Free cancellation')).toBeInTheDocument();
    });

    it('does not show free cancellation when not available', () => {
        render(
            <HotelCard
                {...defaultProps}
                cancellationType="NOT_REFUNDABLE"
            />
        );

        expect(screen.queryByText('Free cancellation')).not.toBeInTheDocument();
    });

    it('shows savings when available', () => {
        render(<HotelCard {...defaultProps} />);

        expect(screen.getByText(/Save \$20.*/)).toBeInTheDocument();
    });

    it('does not show savings when null', () => {
        render(
            <HotelCard
                {...defaultProps}
                savings={null}
            />
        );

        expect(screen.queryByText(/Save/)).not.toBeInTheDocument();
    });

    it('shows promotion tag when available', () => {
        render(
            <HotelCard
                {...defaultProps}
                promotion={{
                    title: 'Special Offer',
                    type: 'MEMBER'
                }}
            />
        );

        expect(screen.getByText('Special Offer')).toBeInTheDocument();
    });
}); 