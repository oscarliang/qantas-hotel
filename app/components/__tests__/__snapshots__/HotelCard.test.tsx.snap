// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelCard Component renders hotel details correctly 1`] = `
<div>
  <div
    class="flex flex-col sm:flex-row  border-b border-gray-200 py-4"
  >
    <div
      class="relative"
    >
      <img
        alt="Test Hotel Image"
        class="object-cover"
        height="125"
        src="/test-image.jpg"
        width="145"
      />
    </div>
    <div
      class="flex-1 px-2 sm:px-4 py-2"
    >
      <div
        class="flex flex-col sm:flex-row justify-between"
      >
        <div
          class="relative min-h-32 sm:min-h-28"
        >
          <div
            class="flex flex-col sm:flex-row justify-content align-middle"
          >
            <h2
              class="text-lg font-bold"
            >
              Test Hotel
            </h2>
            <div
              class="flex my-1 sm:mt-2 sm:ml-2"
            >
              <span
                data-testid="star-icon-filled"
              >
                ★
              </span>
              <span
                data-testid="star-icon-filled"
              >
                ★
              </span>
              <span
                data-testid="star-icon-filled"
              >
                ★
              </span>
              <span
                data-testid="star-icon-filled"
              >
                ★
              </span>
              <span
                data-testid="star-icon-empty"
              >
                ★
              </span>
            </div>
          </div>
          <address
            class="text-sm text-gray-600 not-italic"
          >
            123 Test St, Test City
          </address>
          <p
            class="text-red-600 text-sm mt-2 font-medium underline"
          >
            Deluxe Room
          </p>
          <p
            class="text-sm text-green-600 mt-2 absolute bottom-0"
          >
            Free cancellation
          </p>
        </div>
        <div
          class="text-right mt-4 sm:mt-0"
        >
          <div
            class="text-sm text-gray-600"
          >
            1 night total (AUD)
          </div>
          <div
            class="text-2xl font-bold mt-1"
          >
            <span
              class="text-lg"
            >
              $
            </span>
            200
          </div>
          <div
            class="text-red-600 text-sm"
          >
            Save $
            20
            ~
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
